### PR TITLE
shader warm up with canvaskit and corresponding test

### LINKED ISSUE #75801: [web] we need to warm-up CanvasKit renderer shaders
Currently shader warm-up is conditional behind `kIsWeb` flag because the HTML renderer crashes during warm-up. Because the flag doesn't distinguish between HTML and CanvasKit we end up not warming up CanvasKit either. This leads to jank in the middle of animations after the app starts up. We need a way to do the shader warm-up for the CanvasKit renderer.

### PR DESCRIPTION
**Pull request by alanwutang11**
shader warm up used to be conditional behind kIsWeb flag. Fixed by distinguishing between CanvasKit and HTML with another boolean. 

fixes https://github.com/flutter/flutter/issues/75801
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' as ui;
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+Future<void> main() async {
+
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:11]: nit unnecessary blank line

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  test('ShaderWarmUp', () {
+
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:13]: ditto

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    final FakeShaderWarmUp shaderWarmUp = FakeShaderWarmUp();
+    PaintingBinding.shaderWarmUp = shaderWarmUp;
+    WidgetsFlutterBinding.ensureInitialized();
+    expect(shaderWarmUp.ranWarmUp, true);
+
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:18]: ditto

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+class FakeShaderWarmUp extends ShaderWarmUp {
+
+  bool ranWarmUp = false;
+
+  @override
+  Future<bool> warmUpOnCanvas(ui.Canvas canvas) {
+    ranWarmUp = true;
+    return Future<bool>.delayed(Duration.zero, () => true);
+  }
+
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:31]: ditto

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    final FakeShaderWarmUp shaderWarmUp = FakeShaderWarmUp();
+    PaintingBinding.shaderWarmUp = shaderWarmUp;
+    WidgetsFlutterBinding.ensureInitialized();
+    expect(shaderWarmUp.ranWarmUp, true);
+
+  });
+}
+
+class FakeShaderWarmUp extends ShaderWarmUp {
+
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:23]: ditto

File: packages/flutter/test/painting/shader_warm_up_test.dart [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' as ui;
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+Future<void> main() async {
```
Comment (hterkelsen) [inline packages/flutter/test/painting/shader_warm_up_test.dart:10]: this test should fail when running on web in HTML mode. you need to add a `skip: kIsWeb && !kIsCanvasKit`

**Comment by hterkelsen** [comment]
I just looked at the code more closely and saw that `ShaderWarmUp.warmUpOnCanvas` is called unconditionally. It is only `picture.toImage` on the resulting picture that is gated on `!kIsWeb`: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/shader_warm_up.dart#L90

So what you should do in the test, in addition to what you have already, is set `debugCaptureShaderWarmUpImage` and verify that it is called with a valid, non-null image in the same way as you checked `warmUpOnCanvas`.

**Review by harryterkelsen — APPROVED** [review]
LGTM

**Pull request merged**